src/components/SyncIndicator.tsx

```diff
@@ -0,0 +1,191 @@
+import { next as A } from "@automerge/automerge";
+import { DocHandle } from "@automerge/automerge-repo";
+import { useState, useEffect } from "react";
+import { useRepo } from "@automerge/automerge-repo-react-hooks";
+import { WifiIcon, WifiOffIcon } from "lucide-react";
+
+// if we are connected to the sync server and have outstanding changes we should switch to the error mode if we haven't received a sync message in a while
+// this variable specifies this timeout duration
+const SYNC_TIMEOUT = 3000 
+
+// wait initially if we haven't connnected to the sync server yet
+const INITIAL_CONNECTION_WAIT = 2000
+
+export const SyncIndicator = ({ handle }: { handle: DocHandle<unknown> }) => {
+  // todo: sync shouldn't flicker. It should be based on if we are making progress syncing the changes.
+  // While a user is actively editing the document their heads will never match the synced heads
+  // is is fine and shouldn't be counted as out of sync as long as they are not lagging behind the sync server by too much
+  const {isSynced, lastSyncUpdate} = useSyncStateOfServer(handle);
+  const isOnline = useIsOnline();
+  const isConnectedToServer = useIsConnectedToServer();
+  const hasSyncTimedOut = useHasTimedOut({ 
+    duration: SYNC_TIMEOUT,   
+    timestamp: lastSyncUpdate,
+    isActive: !isSynced
+  })
+  const hasInitialConnectionWaitTimedOut = useHasTimedOut({ duration: INITIAL_CONNECTION_WAIT })
+
+  if (isOnline) {
+    if ((isConnectedToServer || !hasInitialConnectionWaitTimedOut)  && (isSynced || !hasSyncTimedOut)) {
+      return (
+        <div className="text-gray-500">
+          <WifiIcon size={"20px"} className="inline-block mr-[7px]" />
+        </div>
+      );
+    } else {
+      return (
+        <div className="text-red-500 flex items-center">
+          <WifiIcon size={"20px"} className={`inline-block ${isSynced ? "mr-[7px]" : ""}`} />
+          {!isSynced && <div className="inline text-xs">*</div>}
+          <div className="ml-1">Sync Error</div>
+        </div>
+      );
+    }
+  } else {
+    return (
+      <div className="text-gray-500">
+        <WifiOffIcon size={"20px"} className={`inline-block ${isSynced ? "mr-[7px]" : ""}`} />
+        {!isSynced && <div className="inline text-xs">*</div>}
+      </div>
+    );
+  }
+};
+
+const SYNC_SERVER_PREFIX = "storage-server-";
+
+interface TimeoutConfig {
+  duration: number
+  timestamp?: number,
+  isActive?: boolean
```

Comment:
does this basically mean "we're actively checking to see if there's been a timeout"? could you add comments to the fields on this interface